### PR TITLE
DAOS-6566 vos: reject direct key larger than anchor

### PR DESCRIPTION
**Pull request by NiuYawei**
Reject direct key larger than anchor size.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

**Review by jolivier23 — CHANGES_REQUESTED** [review]
I think we should keep the asserts in encode/decode but add an error on insert if a direct key exceeds the limit.  Aggregation will have to iterate them even if the user doesn't so we can't really support them properly unless we fix the anchor protocol

**Review by jolivier23 — DISMISSED** [review]

**Comment by NiuYawei** [comment]
All CI tests passed, but daos_vol tests definitely insert long dkeys.

Server log from daos_vol test:
``
02/05-04:31:45.42 wolf-135 DAOS[45492/45765] tree ERR  src/common/btree.c:483 btr_rec_alloc() Key size (66560) > Anchor size (100)
02/05-04:31:45.42 wolf-135 DAOS[45492/45765] vos  ERR  src/vos/vos_tree.c:1062 key_tree_prepare() Failed to upsert: DER_MISC(-1025): 'Miscellaneous error'
02/05-04:31:45.42 wolf-135 DAOS[45492/45765] vos  ERR  src/vos/vos_io.c:1604 dkey_update() Error preparing dkey tree: rc=DER_MISC(-1025): 'Miscellaneous error'
02/05-04:31:45.42 wolf-135 DAOS[45492/45765] vos  ERR  src/vos/vos_io.c:2072 vos_update_end() Failed to update tree index: DER_MISC(-1025): 'Miscellaneous error'
``

Client test log from daos_vol test:
``
2021-02-05 04:31:45,634 process          L0479 DEBUG| [stdout] Testing soft link creation with a long name                           *FAILED*
2021-02-05 04:31:45,634 process          L0479 DEBUG| [stderr]   #000: /builddir/build/BUILD/vol-daos-1.1.0rc2/src/daos_vol_link.c line 2348 in H5_daos_link_create(): link creation failed in task "link write": miscellaneous error (DER_MISC)
2021-02-05 04:31:45,634 process          L0479 DEBUG| [stdout]     couldn't create soft link with a long name
2021-02-05 04:31:45,634 process          L0479 DEBUG| [stderr]     major: Links
2021-02-05 04:31:45,635 process          L0479 DEBUG| [stderr]     minor: Can't operate on object
2021-02-05 04:31:45,635 process          L0479 DEBUG| [stderr]   #001: /builddir/build/BUILD/vol-daos-1.1.0rc2/src/daos_vol.c line 3287 in H5_daos_h5op_finalize(): operation "soft link create" failed in task "link write": miscellaneous error (DER_MISC)
``

But the daos_vol was reported as PASS despite of all these failures.

I think we'd land this fix to avoid long direct dkey inserted, otherwise, it could trigger the assert on client enumeration or server aggregation.

**Review by jolivier23 — DISMISSED** [review]

**Comment by mchaarawi** [comment]
this is weird, because on my side i haven't seen those tests failing before.
in latest master run, they work fine:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/3781/artifact/Functional/job-2021-02-08T06.32-fe676d5-io-daos_vol/job.log

aren't they supposed to cause server assertion or fail??
i don't think we should land this PR TBH before we understand and fix the issue.

**Comment by NiuYawei** [comment]
> this is weird, because on my side i haven't seen those tests failing before.
> in latest master run, they work fine:
> https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/3781/artifact/Functional/job-2021-02-08T06.32-fe676d5-io-daos_vol/job.log
> 
> aren't they supposed to cause server assertion or fail??
> i don't think we should land this PR TBH before we understand and fix the issue.

@mchaarawi , in latest master, we don't reject long direct key, so you won't see test failures, and the daos_vol test always finish quickly (container is destroyed) before aggregation is triggered, so we don't observe the server assertion as well (it happens on VOS tree iteration only, like aggregation or processing enumeration request from client)

In one of my PR I hit the server assertion quite often, because there was a defect in my PR caused delay on container close, so the aggregation was triggered in daos_vol test and assertion was triggered.

**Comment by mchaarawi** [comment]
> > this is weird, because on my side i haven't seen those tests failing before.
> > in latest master run, they work fine:
> > https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/3781/artifact/Functional/job-2021-02-08T06.32-fe676d5-io-daos_vol/job.log
> > aren't they supposed to cause server assertion or fail??
> > i don't think we should land this PR TBH before we understand and fix the issue.
> 
> @mchaarawi , in latest master, we don't reject long direct key, so you won't see test failures, and the daos_vol test always finish quickly (container is destroyed) before aggregation is triggered, so we don't observe the server assertion as well (it happens on VOS tree iteration only, like aggregation or processing enumeration request from client)
> 
> In one of my PR I hit the server assertion quite often, because there was a defect in my PR caused delay on container close, so the aggregation was triggered in daos_vol test and assertion was triggered.

yes I had a short chat with Jeff yesterday to explain this issue. i think we need to modify the VOL tests to shorten this key entry they write and rebuild the RPMs for that before we can land this patch though.

**Comment by NiuYawei** [comment]
> > > this is weird, because on my side i haven't seen those tests failing before.
> > > in latest master run, they work fine:
> > > https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/3781/artifact/Functional/job-2021-02-08T06.32-fe676d5-io-daos_vol/job.log
> > > aren't they supposed to cause server assertion or fail??
> > > i don't think we should land this PR TBH before we understand and fix the issue.
> > 
> > 
> > @mchaarawi , in latest master, we don't reject long direct key, so you won't see test failures, and the daos_vol test always finish quickly (container is destroyed) before aggregation is triggered, so we don't observe the server assertion as well (it happens on VOS tree iteration only, like aggregation or processing enumeration request from client)
> > In one of my PR I hit the server assertion quite often, because there was a defect in my PR caused delay on container close, so the aggregation was triggered in daos_vol test and assertion was triggered.
> 
> yes I had a short chat with Jeff yesterday to explain this issue. i think we need to modify the VOL tests to shorten this key entry they write and rebuild the RPMs for that before we can land this patch though.

I didn't find the daos_vol test code, these tests are maintained by us or other groups?

**Review by gnailzenh — APPROVED** [review]

File: src/common/btree.c [hunk 1]
```diff
@@ -477,6 +477,14 @@ static int
 btr_rec_alloc(struct btr_context *tcx, d_iov_t *key, d_iov_t *val,
 	       struct btr_record *rec)
 {
+	if (btr_is_direct_key(tcx)) {
+		if (key->iov_len > EMBEDDED_KEY_MAX) {
+			D_ERROR("Key size (%zd) > Anchor size (%u)\n",
+				key->iov_len, EMBEDDED_KEY_MAX);
+			return DER_KEY2BIG;
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/common/btree.c:484]: indeed.

File: src/common/btree.c [hunk 2]
```diff
@@ -477,6 +477,14 @@ static int
 btr_rec_alloc(struct btr_context *tcx, d_iov_t *key, d_iov_t *val,
 	       struct btr_record *rec)
 {
+	if (btr_is_direct_key(tcx)) {
+		if (key->iov_len > EMBEDDED_KEY_MAX) {
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/common/btree.c:481]: ok

**Review by jolivier23 — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**